File: BinaryTrees/PathSum.java
package BinaryTrees;

import java.util.Stack;

/*Given the root of a binary tree and an integer targetSum, 
return true if the tree has a root-to-leaf path such that adding up all the values along the path equals targetSum.

A leaf is a node with no children. */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class PathSum {
    public boolean hasPathSumRecursive(TreeNode root, int targetSum) {
        //return false if tree/subtree is empty
        if (root == null) {
            return false;
        }

        //subtract node's value from target sum
        targetSum -= root.val;

        //if node is leaf, see if targetSum now equals 0 (the sum of path would equal target)
        if (root.left == null && root.right == null) {
            return targetSum == 0;
        }

        //return true if left or right contains path to equaling target
        return hasPathSumRecursive(root.left, targetSum) || hasPathSumRecursive(root.right, targetSum);
    }

    public boolean hasPathSumIterative(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }

        Stack<Pair> stack = new Stack<>();
        stack.push(new Pair(root, 0));
        
        while (!stack.isEmpty()) {
            Pair pair = stack.pop();
            TreeNode currentNode = pair.node;
            int currentSum = pair.previousSum + currentNode.val;

            //if current node is leaf, return true if sum equals target
            if (currentNode.left == null && currentNode.right == null) {
                if (currentSum == targetSum) return true;
            }
            
            //push non-null children to stack to keep evaluating path
            if (currentNode.left != null) stack.push(new Pair(currentNode.left, currentSum));
            if (currentNode.right != null) stack.push(new Pair(currentNode.right, currentSum));
            
        }

        //if stack empties and no valid path is found, return false
        return false;
    }
}

class Pair { 
    TreeNode node;
    int previousSum;
    Pair(TreeNode node, int previousSum) {
        this.node = node;
        this.previousSum = previousSum;
    }
}

File: BinaryTrees/path_sum.py
from collections import defaultdict
from typing import List, Optional

from BinaryTrees.tree_node import TreeNode


class PathSum:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        def dfs(node: Optional[TreeNode], prev_vals: List[int]) -> int:
            if not node:
                return 0

            count = 0 if node.val != targetSum else 1
            curr_sum = node.val
            for i in range(len(prev_vals) -1, -1, -1):
                val = prev_vals[i]
                curr_sum += val
                if curr_sum == targetSum:
                    count += 1
            prev_vals.append(node.val)
            count += dfs(node.left, prev_vals)
            count += dfs(node.right, prev_vals)
            prev_vals.pop()
            return count

        return dfs(root, [])

    def pathSumOptimized(self, root: Optional[TreeNode], targetSum: int) -> int:
        def dfs(node: Optional[TreeNode], sum_map: dict[int, int], curr_sum) -> int:
            if not node:
                return 0

            count = 0
            curr_sum += node.val
            count += sum_map[curr_sum - targetSum]
            sum_map[curr_sum] += 1
            count += dfs(node.left, sum_map, curr_sum)
            count += dfs(node.right, sum_map, curr_sum)
            sum_map[curr_sum] -= 1
            return count

        sum_map = defaultdict(int)
        sum_map[0] = 1
        return dfs(root, sum_map, 0)
